### ISSUE TITLE
Proposal: End support Windows versions prior to Windows 10

### ISSUE THREAD

**Issue by teamterraform** [hashicorp/terraform#22487]
Terraform currently has a number of special compatibility codepaths to help it work correctly with older Windows versions.

The most significant of these is the translation layer between the VT100-style escape sequences that Terraform produces to the Windows Console API. The Windows Console API in older Windows versions has some significant differences to a VT100 terminal:
- Windows Console expects background and foreground color attributes to be set together, whereas VT100 allows the foreground color to be changed alone. This is why on Windows in a Console window with its background color set to anything other than black the non-default colored text from Terraform appears against a black background.
- Windows Console does not support formatting attributes such as bold and underline. We can simulate bold by using a more intense color, but underline is not available at all.
- Although the Windows Console _does_ support writing characters to locations other than the current cursor position and changing the cursor position programmatically, the translation layer we currently use does not translate the corresponding VT100 sequences consistently enough with VT100-compatible terminal behavior for us to confidently use techniques such as in-place updating of progress indicators, etc.
- Until the Windows 10 October 2018 update, the Console window could render only characters from the current ANSI character set, rather than the full complement of Unicode.

The Windows Console subsystem has been [significantly redesigned](https://devblogs.microsoft.com/commandline/windows-command-line-inside-the-windows-console/) in Windows 10 updates over the last year or so, motivated by the new ability to run Linux-oriented applications through the Console using [Windows Subsystem for Linux](https://en.wikipedia.org/wiki/Windows_Subsystem_for_Linux) (WSL). It now has a comparable featureset to a virtual terminal, which could allow us to remove the Console API translation layer and use VT100 sequences directly on Windows, if we are able to end support for older Windows versions.

That should in turn allow us more flexibility in improving the Terraform CLI user experience. An immediate consequence of removing the translation layer would be that colored text would no longer be forced to render with a black background, underlines in diagnostic messages would be visible, and non-ASCII strings in configuration should render reliably when displayed in Terraform's output.

In the longer term, it could allow for more compact and concise output through updated-in-place progress indicators and other similar dynamic elements, which we've generally avoided so far to minimize cross-platform inconsistencies and bugs.

---

This is an exploratory issue to investigate what, if anything, blocks removing support for Windows 9, and thus whether it would be feasible to do so in a forthcoming Terraform release.

Because we ship a single distribution of Terraform targeting Windows, we don't currently have any signal on which Windows versions that distribution is being used on, only how much that version is being downloaded in relation to others. It's likely that we will need to conduct some specific user research in order to collect more data to proceed here. No immediate action will occur as a result of this issue; this issue is representing the need to devise a plan, not to execute on that plan.

For now, this issue is a placeholder for the idea of ending legacy Windows support. If you are currently running Terraform on Windows 9 or earlier, or on any Windows 10 (or Windows Server 2019) release whose Console implementation lacks the modern VT features, we'd love to hear about what you are using it for and what is blocking you from upgrading to a more modern Windows release.

**Comment by phil-bevan** [comment]
Personally I'm running Windows 10 so am happy for you to do what you like!  However, some observations:
1.  There is no such thing as Windows 9.  Windows versions skipped from 8 to 10
2.  Removing support for any version of Windows that is still considered in mainstream support by Microsoft is going to be problematic for many of your customers who are in large orgs.  These orgs upgrade at a glacial pace, frequently skipping OS versions to reduce the frequency (and cost, and risk) of upgrades.  Windows 7 is just going out of support, so most Win7 customers will have jumped to Win10, but you will have a large number of Win8 orgs who will not move until Windows 11 releases.

**Comment by Dogers** [comment]
Just an update on the above..

* Windows 7 is EOL since the beginning of the year, but still has a ~15-20% share. While there's an argument to continue supporting it, if the vendor doesn't then why should anyone else? I don't know the breakdown of this between enterprise and home though
* Windows 8 is EOL for a while now
* Windows 8.1 is still in support until 2023, but has ~4% usage
* Windows Terminal (which I assume this is what it's all really about!) needs version 1903, but the prior version is still in support until mid May 2021

**Comment by apparentlymart** [comment]
Hi all!

For those who have been watching this issue, I wanted to add a note about the compromise we eventually reached after some research and development.

Terraform v0.15 and later will remove the compatibility layer we were previously using to translate on the fly from "xterm-style" virtual terminal sequences into the out-of-band classic Windows Console API. We made this decision after reviewing Microsoft's advice in [Classic Console APIs versus Virtual Terminal Sequences](https://docs.microsoft.com/en-us/windows/console/classic-vs-vt), and after testing how this new behavior would degrade if used on Windows 8 or on early versions of Windows 10 that predate the addition of the VT mode in the console API.

The high-level position is that we're now only _officially_ supporting Terraform on versions of Windows which support virtual terminal mode in the console, but we've made an effort to verify that this initial release remains generally usable on Windows 8.1, albeit requiring one of the following workarounds in order to avoid visible control sequences in the output:

* Use Terraform's existing `-no-color` option to disable the formatting sequences altogether. Terraform will then produce non-colored output, but should otherwise behave correctly.
* Use a third-party virtual terminal utility to run Terraform, instead of the built-in Windows console. Some example packages are [ConEmu](https://conemu.github.io/), [Cmder](https://cmder.net/), and [mintty](https://mintty.github.io/).

Separately from the virtual terminal sequences, Terraform has also typically expected the terminal to accept the UTF-8 character encoding. Classing Windows Console does not, but we historically reduced the impact of that by constraining Terraform's own output to only include ASCII characters, and thus non-ASCII characters would appear in the output only if given as part of user input such as in the configuration. Since all versions of Windows which support virtual terminal also support the new UTF-8-compatible console buffer, future versions of Terraform will increasingly make use of a fuller range of Unicode characters, such as box drawing characters, which can appear as corrupted in the Windows console on older versions of Windows where the console buffer does not fully support Unicode.

While we hope that this best effort to keep Terraform v0.15.0 usable on Windows 8 will make it easier to plan upgrade strategy, we don't have the resources to proactively test on Windows 8 on an ongoing basis and therefore we can't promise that future Terraform releases will remain Windows 8 compatible. Of course, prior releases of Terraform CLI remain available for download, so you might prefer to just remain on your current version of Terraform until you are ready to upgrade Windows.

Thanks to those who left comments here with various details about what was added in which version of Windows and which versions of Windows remain supported at what level.
